### PR TITLE
Quiz: augment quiz questions

### PR DESCRIPTION
**Pull request by robinmessage**
---

**Pull Request Check List**
- [ ] Added enough Logging to monitor expected behaviour change
- [x] Security - Injection - everything run by an interpreter (SQL, OS...) is either validated or escaped
- [x] Security - Data Exposure - PII is not stored or sent unencrypted
- [x] Security - Data Exposure - Test any altered or created endpoints using [swagger](http://localhost:8080/isaac-api/api-docs/)
- [x] Security - Access Control - Check authorisation on every new endpoint
- [ ] DB schema changes - postgres-rutherford-create-script updated
- [ ] DB schema changes - upgrade script created matching create script
- [ ] Updated Release Procedure & Documentation (& Considered Implications to Previous Versions)
- [ ] Peer-Reviewed

### CODE REVIEW

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/managers/QuizQuestionManager.java [hunk 1]
```diff
@@ -81,4 +85,56 @@ public void recordQuestionAttempt(QuizAttemptDTO quizAttempt, QuestionValidation
...
+        Map<String, List<QuestionValidationResponse>> answers = quizQuestionAttemptManager.getAnswers(quizAttempt.getId());
+
+        for (QuestionDTO question : questionsToAugment) {
+            List<QuestionValidationResponse> questionAttempts = answers.get(question.getId());
+
+            // The latest answer is the only answer we consider.
+            if (questionAttempts.size() > 0) {
+                QuestionValidationResponseDTO lastAttempt = questionManager.convertQuestionValidationResponseToDTO(questionAttempts.get(questionAttempts.size() - 1));
+                if (!includeCorrect) {
+                    lastAttempt.setCorrect(null);
```
Comment (jsharkey13) [inline src/main/java/uk/ac/cam/cl/dtg/isaac/api/managers/QuizQuestionManager.java:134]: Will this not leave the `explanation` field from the validation response still present, which might explain the correct answer or inform them of the correct/incorrectness. What about `QuantityValidationResponseDTO` s which have `correctValue` and `correctUnits`? (There are other subclasses, but that is the only one that is exposed to the user other than the base class by `QuestionValidationResponseOrikaConverter`).

I wonder whether the safest thing to do is to manually create a new `QuestionValidationResponseDTO`, with only the `questionId`, `answer` and `dateAttempted` fields copied over from the actual DO, using a new private method?

File: src/main/java/uk/ac/cam/cl/dtg/isaac/dao/PgQuizQuestionAttemptPersistenceManager.java [hunk 2]
```diff
@@ -75,4 +81,35 @@ public void registerQuestionAttempt(Long quizAttemptId, QuestionValidationRespon
             throw new SegueDatabaseException("Unable to process json exception", e);
         }
     }
+
+    @Override
+    public Map<String, List<QuestionValidationResponse>> getAnswers(Long quizAttemptId) throws SegueDatabaseException {
```
Comment (jsharkey13) [inline src/main/java/uk/ac/cam/cl/dtg/isaac/dao/PgQuizQuestionAttemptPersistenceManager.java:86]: Could this be clearer, `getAllAnswersForQuizAttempt` or something?

**Review by jsharkey13 — CHANGES_REQUESTED** [review]
I think this will leak the correctness of answers, despite trying not to. Our validation response classes are a bit messy, and just blanking out `correct` isn't going to be enough I don't think!

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/managers/QuizQuestionManager.java (hunk 1 above)
Comment (robinmessage) [reply src/main/java/uk/ac/cam/cl/dtg/isaac/api/managers/QuizQuestionManager.java:134]: Yes, I think I added `lastAttempt.setExplanation(null);` but maybe that was on another PR?

In any case, I hadn't thought about subclasses. I think manually copying out of the DO into a DTO with those fields would be safer, yes. I'll do that.

File: src/main/java/uk/ac/cam/cl/dtg/isaac/dao/PgQuizQuestionAttemptPersistenceManager.java (hunk 2 above)
Comment (robinmessage) [reply src/main/java/uk/ac/cam/cl/dtg/isaac/dao/PgQuizQuestionAttemptPersistenceManager.java:86]: Sure, can do.

**Pull request merged**